[SPARK-13267] [Web UI] document the ?param arguments of the REST API; lift the…

steveloughran opened the pull request:
Add to the REST API details on the ? args and examples from the test suite.

I've used the existing table, adding all the fields to the second table.

see [in the pr](https://github.com/steveloughran/spark/blob/history/SPARK-13267-doc-params/docs/monitoring.md).

There's a slightly more sophisticated option: make the table 3 columns wide, and for all existing entries, have the initial `td` span 2 columns. The new entries would then have an empty 1st column, param in 2nd and text in 3rd, with any examples after a `br` entry.

steveloughran committed:
[SPARK-13267] document the ?param arguments of the REST API; lift the examples from the test suite

SparkQA commented:
**[Test build #51037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51037/consoleFull)** for PR 11152 at commit [`bf1a128`](https://github.com/apache/spark/commit/bf1a128de6f2af3f7bba2aef6d546fcca1aaa19b).

SparkQA commented:
**[Test build #51037 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51037/consoleFull)** for PR 11152 at commit [`bf1a128`](https://github.com/apache/spark/commit/bf1a128de6f2af3f7bba2aef6d546fcca1aaa19b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51037/
Test PASSed.

squito commented:
Here's a screenshot of what this looks like:

<img width="929" alt="screen shot 2016-02-11 at 10 44 56 am" src="https://cloud.githubusercontent.com/assets/71240/12982980/e01be154-d0ac-11e5-94b1-3921c24298e5.png">

Its not wonderful, but seems pretty reasonable without some bigger restructuring of the page.  (Eg. as one random example, see this: https://docs.atlassian.com/jira/REST/latest/#api/2/-getPermissions  But that is clearly a bigger change.)

squito commented on docs/monitoring.md:
```diff
@@ -189,10 +189,26 @@ for a running application, at `http://localhost:4040/api/v1`.
+  </tr>
+  <tr>
+     <td></td>
+     <td><code>?maxDate=[date]</code> latest date to list. e.g <code>?maxDate=2015-02-03T16:42:40.000GMT</code></td>
```
maybe mention both date formats under `minDate`, and then just say this can take the same formats as `minDate`?  It should be pretty obvious that both versions work for either option, but just to be a little more clear.

squito commented:
lgtm other than one minor comment

steveloughran committed:
SPARK-13267 take on suggested changes in layout

steveloughran commented:
done. Also put the examples on new lines with a &lt;br;&gt; tag

steveloughran commented:
<table class="table">
  <tr><th>Endpoint</th><th>Meaning</th></tr>
  <tr>
    <td><code>/applications</code></td>
    <td>A list of all applications</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?status=[completed|running]</code> list only applications in the chosen state</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?minDate=[date]</code> earliest date/time to list. Examples:
     <br><code>?minDate=2015-02-10</code>
     <br><code>?minDate=2015-02-03T16:42:40.000GMT</code></td>
  </tr>
  <tr>
     <td></td>
     <td><code>?maxDate=[date]</code> latest date/time to list; uses same format as <code>minDate</code></td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/jobs</code></td>
    <td>A list of all jobs for a given application</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?status=[complete|succeeded|failed]</code> list only jobs in the state</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/jobs/[job-id]</code></td>
    <td>Details for the given job</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/stages</code></td>
    <td>A list of all stages for a given application</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/stages/[stage-id]</code></td>
    <td>A list of all attempts for the given stage</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?status=[active|complete|pending|failed]</code> list only stages in the state</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/stages/[stage-id]/[stage-attempt-id]</code></td>
    <td>Details for the given stage attempt</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/stages/[stage-id]/[stage-attempt-id]/taskSummary</code></td>
    <td>Summary metrics of all tasks in the given stage attempt</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?quantiles</code> summarize the metrics with the given quantiles.
     Example: <br><code>?quantiles=0.01,0.5,0.99</code></td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/stages/[stage-id]/[stage-attempt-id]/taskList</code></td>
    <td>A list of all tasks for the given stage attempt</td>
  </tr>
  <tr>
     <td></td>
     <td><code>?offset=[offset]&amp;length=[len]</code> list only tasks in the range.
      Example: <br> <code>?offset=10&amp;length=50</code></td>
  </tr>
  <tr>
     <td></td>
     <td><code>?sortBy=[runtime|-runtime]</code> sort the tasks</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/executors</code></td>
    <td>A list of all executors for the given application</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/storage/rdd</code></td>
    <td>A list of stored RDDs for the given application</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/storage/rdd/[rdd-id]</code></td>
    <td>Details for the storage status of a given RDD</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/logs</code></td>
    <td>Download the event logs for all attempts of the given application as a zip file</td>
  </tr>
  <tr>
    <td><code>/applications/[app-id]/[attempt-id]/logs</code></td>
    <td>Download the event logs for the specified attempt of the given application as a zip file</td>
  </tr>
</table>

SparkQA commented:
**[Test build #51190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51190/consoleFull)** for PR 11152 at commit [`15c675d`](https://github.com/apache/spark/commit/15c675d8b9fbfaa47e2915ad0075ebe03f1b5bcf).

SparkQA commented:
**[Test build #51190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51190/consoleFull)** for PR 11152 at commit [`15c675d`](https://github.com/apache/spark/commit/15c675d8b9fbfaa47e2915ad0075ebe03f1b5bcf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51190/
Test PASSed.

steveloughran committed:
[SPARK-13267] paah on logs

SparkQA commented:
**[Test build #52754 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52754/consoleFull)** for PR 11152 at commit [`e4e5ae8`](https://github.com/apache/spark/commit/e4e5ae8bcde2fffb00752c71592ff049eb833e04).

SparkQA commented:
**[Test build #52754 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52754/consoleFull)** for PR 11152 at commit [`e4e5ae8`](https://github.com/apache/spark/commit/e4e5ae8bcde2fffb00752c71592ff049eb833e04).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52754/
Test PASSed.

steveloughran commented:
@squito : think this is ready for a merge now. I've been testing code against it and not found any inconsistencies

squito commented on docs/monitoring.md:
```diff
@@ -273,8 +309,8 @@ for a running application, at `http://localhost:4040/api/v1`.
-    <td><code>/applications/[app-id]/[attempt-id]/logs</code></td>
-    <td>Download the event logs for the specified attempt of the given application as a zip file</td>
+    <td><code>/applications/[app-id]/logs</code></td>
+    <td>Download the event logs for the application as a zip file</td>
```
this change looks wrong -- there are two endpoints, one for all attempts for the applications, and another for just one attempt.  The docs for the all-attempt endpoint is above this one.  with this change, its just duplicated.

steveloughran replied:
OK, now i'm confusing, because the docs at the bottom say appId is `appid/attemptid` in yarn, which is true for every other element, which was why I adjusted this one. 

I'll try and clarify things more about appId vs attemptId, pull the details to the top and then have special mention here. As clearly it managed to confuse me.

squito replied:
ah good point.  So this case is a little different because both are supported and the one w/out `attemptId` is sometimes useful, to grab the logs for all attempts.  I thought the docs do actually spell this out, the first endpoint says it will download logs for all attempts.

squito commented on docs/monitoring.md:
```diff
@@ -229,10 +229,28 @@ for a running application, at `http://localhost:4040/api/v1`.
+  <tr>
+     <td></td>
+     <td><code>?minDate=[date]</code> earliest date/time to list. Examples:
+     <br><code>?minDate=2015-02-10</code>
```
super-nit: should these be `<br/>`?  I guess no browsers care about the `/` anymore, but might as well ...

steveloughran replied:
actually no, it MUST be `<br>` without a tag, except in the specific case of XHTML, which nobody uses

1. [Mozilla](https://developer.mozilla.org/en/docs/Web/HTML/Element/br)
1. [MSFT](https://msdn.microsoft.com/en-us/library/ms535208(v=vs.85).aspx)

This is the same as the paragraph tag, {{<p>}}.

steveloughran committed:
[SPARK-13267] pull up details on app-id on YARN; pull out base-app-id for the special case on logs. Fix some indentations

steveloughran commented:
Fixed up the endpoints, added some more detail on app-id vs app-attempt, using base-app-id for the log retrieval. Also mentioned that after job/stage GC, there are less to retrieve, and that you can increase the number and restart the HS, if you really want more

SparkQA commented:
**[Test build #52927 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52927/consoleFull)** for PR 11152 at commit [`2895a4f`](https://github.com/apache/spark/commit/2895a4f3da0cd9cc30e1db3dc1a7fdf00eff2149).

SparkQA commented:
**[Test build #52927 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52927/consoleFull)** for PR 11152 at commit [`2895a4f`](https://github.com/apache/spark/commit/2895a4f3da0cd9cc30e1db3dc1a7fdf00eff2149).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52927/
Test PASSed.

steveloughran commented:
@squito have you had a chance to review this?

srowen commented on docs/monitoring.md:
```diff
@@ -269,17 +310,23 @@ for a running application, at `http://localhost:4040/api/v1`.
-    <td><code>/applications/[app-id]/[attempt-id]/logs</code></td>
-    <td>Download the event logs for the specified attempt of the given application as a zip file</td>
+    <td><code>/applications/[base-app-id]/[attempt-id]/logs</code></td>
+    <td>Download the event logs for a specific application as a zip file</td>
```
Is this right, because in this form the attempt ID is specified?

squito replied:
I think the idea is that in all other cases `[app-id]` *must* include the attempt-id when applicable as well.  However `/logs` is special, in that even if there are multiple attempts, there is still an end-point with just base-app-id for getting logs from all attempts.

steveloughran replied:
Looking a the the implementation in `FsHistoryProvider`, it states " If no attempt is specified, or there is no attemptId for attempts, return all attempts"

That is: the zip file contains the logs of *every* attempt. I'll add that to the text.

srowen commented on docs/monitoring.md:
```diff
@@ -222,17 +222,40 @@ both running applications, and in the history server.  The endpoints are mounted
+  </tr>
+  <tr>
+    <td></td>
+    <td><code>?minDate=[date]</code> earliest date/time to list. Examples:
```
Formatting-wise, nit, can this be accomplished with `<p>` or `<br/>` in one big `<td>` rather than make extraneous rows, or does it just not look nice?

srowen commented on docs/monitoring.md:
```diff
@@ -222,17 +222,40 @@ both running applications, and in the history server.  The endpoints are mounted
   </tr>
   <tr>
+    <td></td>
+    <td><code>?status=[completed|running]</code> list only applications in the chosen state</td>
```
And just sense-checking -- all these params are definitely already supported in 2.0, you're just documenting them?

squito replied:
yes, no changes to this in 2.0

squito commented:
Sorry I had forgotten about this.  Good question about just using line breaks instead of separate rows.  Here is a screen shot with separate rows:

<img width="951" alt="separate_rows" src="https://cloud.githubusercontent.com/assets/71240/14611159/7d8bcb4c-0556-11e6-9b88-d52fb550786d.png">

And here it is just with linebreaks:

<img width="940" alt="one_row_linebreaks" src="https://cloud.githubusercontent.com/assets/71240/14611164/854a7fa4-0556-11e6-8e71-e00454216684.png">

I don't really have a strong preference, so think maybe we should just go with `<br>` for simplicity?

steveloughran commented:
OK, I'll move to {{&lt;br>}} tags

steveloughran commented:
Here's a couple of screenshots of two MD editors: IntelliJ IDEA and Mou. Both of these are placing the /applications text in the centre of all the text, which really is ugly

That may not happen in github renderings, but its clear that some MD processors are reaching different conclusions about vertical text alignments.

To avoid that the table will need either {{valign=top}} or the td elements retained. Those two IDEs do take valign happily; I'll look at how the site builds go with <br> & valign.

![screen shot 2016-04-18 at 21 12 07](https://cloud.githubusercontent.com/assets/162090/14618387/4c71ef56-05aa-11e6-9ece-5173c2083a6a.png)


![screen shot 2016-04-18 at 21 12 13](https://cloud.githubusercontent.com/assets/162090/14618388/4c73e70c-05aa-11e6-91dd-aec59bb3b279.png)

steveloughran commented:
OK: for the jekyll site all is well with and without vtable; its just the dev tools where things go wrong. I'll push up to github to see what decisions it makes. Note that the markdown editors both centre all the options in the spark configuration options table —I could add a valign there tag too

steveloughran committed:
[SPARK-13267] put applications/ API entry on a row with br tags after

SparkQA commented:
**[Test build #56132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56132/consoleFull)** for PR 11152 at commit [`d609c89`](https://github.com/apache/spark/commit/d609c8914f65740628fc210e77ca76a73e73b97b).

SparkQA commented:
**[Test build #56132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56132/consoleFull)** for PR 11152 at commit [`d609c89`](https://github.com/apache/spark/commit/d609c8914f65740628fc210e77ca76a73e73b97b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56132/
Test PASSed.

srowen commented on docs/monitoring.md:
```diff
@@ -269,17 +303,24 @@ for a running application, at `http://localhost:4040/api/v1`.
 
-When running on Yarn, each application has multiple attempts, so `[app-id]` is actually
-`[app-id]/[attempt-id]` in all cases.
+The number of jobs and stages which can retrieved is constrained by the same garbage
```
I don't really mind the text here but it doesn't actually trigger GC right? this is just about how much stuff is retained in memory?

steveloughran replied:
true, it's a sort of meta-level GC: go over the limit and the UI just throws away things. How about:  "the retention policy of the standalone spark UI"

srowen commented on docs/monitoring.md:
```diff
@@ -253,10 +273,24 @@ for a running application, at `http://localhost:4040/api/v1`.
   <tr>
+    <td></td>
+    <td><code>?quantiles</code> summarize the metrics with the given quantiles.
+    Example: <br><code>?quantiles=0.01,0.5,0.99</code></td>
```
Technically `<br/>` but no big deal

steveloughran replied:
Now that's interesting. in HTML 1.0-5.x it's like {{<p>}}: you don't insert a closing element in , though its permitted. See: [The br element](http://w3c.github.io/html/textlevel-semantics.html#the-br-element).

If the output goes straight to HTML then, leaving it open is the correct action. 

If, alternatively, its going to go through a process pipeline that generates XML and then things from it, then the closing tag matters.

What is the pipeline that jekyll implements?

srowen replied:
Whatever works, I thought HTML was technically supposed to act like XML and use the self-closing syntax for tags with no body.

steveloughran replied:
In fact it is worse than that: the java8 javadoc engine will fail if a javadoc has an "illegal" closing tag in it. Which means that putting the closing tags in there breaks builds on java 8+. That's why I'm fairly cautious about closing the "classic" HTML tags these days.

srowen commented on docs/monitoring.md:
```diff
@@ -222,17 +222,33 @@ both running applications, and in the history server.  The endpoints are mounted
   </tr>
   <tr>
+    <td></td>
+    <td><code>?status=[complete|succeeded|failed]</code> list only jobs in the state</td>
```
I do think the form with one `<td>` looks meaningfully better. Otherwise these looks like hanging docs for some unspecified endpoint

steveloughran replied:
OK, I'll go with that and `<br/>`

steveloughran committed:
[SPARK-13267] switch to <br> tags in formatting

SparkQA commented:
**[Test build #56385 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56385/consoleFull)** for PR 11152 at commit [`1ec238f`](https://github.com/apache/spark/commit/1ec238ff12c6a7b47cc4aa0993104cba5098df5c).

SparkQA commented:
**[Test build #56385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56385/consoleFull)** for PR 11152 at commit [`1ec238f`](https://github.com/apache/spark/commit/1ec238ff12c6a7b47cc4aa0993104cba5098df5c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56385/
Test PASSed.

srowen commented:
Merged to master

steveloughran commented:
thanks for reviewing this, 
Steve
